### PR TITLE
generation of  hexadecimal floating point numbers in tests

### PR DESCRIPTION
**Pull request by mgudemann**
removed an accidental change, added code to process special cases: NaN and
infinity made required simplications, made corrections in treatment of special
cases

supersedes https://github.com/diffblue/cbmc/pull/1061

### CODE REVIEW

File: src/java_bytecode/expr2java.cpp [hunk 1]
```diff
@@ -235,6 +237,63 @@ std::string expr2javat::convert_constant(
...
+  {
+    ieee_floatt ieee_repr(to_constant_expr(src));
+    std::string result;
+
+    bool float_flag=(src.type()==java_float_type());
+    if(ieee_repr.is_zero())
+    {
+      result+='0';
+      format_spect fs;
+      if(fs.precision>0)
```
Comment (smowton) [inline src/java_bytecode/expr2java.cpp]: This is a bit odd -- either factor out the default if that's what you intend, or initialise this from somewhere

File: src/java_bytecode/expr2java.cpp [hunk 2]
```diff
@@ -235,6 +237,63 @@ std::string expr2javat::convert_constant(
     dest+='L';
     return dest;
   }
+  else if((src.type()==java_float_type()) ||
+          (src.type()==java_double_type()))
+  {
+    ieee_floatt ieee_repr(to_constant_expr(src));
+    std::string result;
+
+    bool float_flag=(src.type()==java_float_type());
```
Comment (smowton) [inline src/java_bytecode/expr2java.cpp]: Rename `is_java_float`

File: src/java_bytecode/expr2java.cpp (hunk 1 above)
Comment (smowton) [inline src/java_bytecode/expr2java.cpp]: Thinking further, I suspect we want 0.0 at all times here? We want an expression that is unambiguously zero of the given type; I don't see that 0.000000 (by default init of format_spect) is any better.

Comment (kroening) [inline src/java_bytecode/expr2java.cpp]: Yes, 0.0 is enough; no point to add further trailing zeros.

However, note that there's -0.0 and 0.0, and that they are different.

File: src/java_bytecode/expr2java.cpp [hunk 3]
```diff
@@ -235,6 +237,63 @@ std::string expr2javat::convert_constant(
...
+      format_spect fs;
+      if(fs.precision>0)
+      {
+        result+='.';
+        for(std::size_t i=0; i<fs.precision; i++)
+          result+='0';
+        if(float_flag)
+          result+='f';
+      }
+      return(result);
```
Comment (kroening) [inline src/java_bytecode/expr2java.cpp]: No () around the thing that's being returned.

File: src/java_bytecode/expr2java.cpp (hunk 2 above)
Comment (mgudemann) [inline src/java_bytecode/expr2java.cpp]: done

File: src/java_bytecode/expr2java.cpp (hunk 1 above)
Comment (mgudemann) [inline src/java_bytecode/expr2java.cpp]: done, outputs `-0.0` in case of `is_zero()` and `get_sign()`

File: src/java_bytecode/expr2java.cpp (hunk 3 above)
Comment (mgudemann) [inline src/java_bytecode/expr2java.cpp]: done

**Comment by smowton** [comment]
Looks better :) Please add regression tests for each of the special cases mentioned here -- NaN, infinities, negative zero and so on.

**Comment by mgudemann** [comment]
@smowton the regression tests are in diffblue/test-gen#909
